### PR TITLE
Get function invocation statistics/metrics from Application Insights

### PR DESCRIPTION
**Pull request by imranalisyed506**
### Problem Description
Get function invocation statistics/metrics from Application Insights

### Solution Description
Implemented getting azure function invocation statistics from Application Insights as optional when Application insights is configured.

### CODE REVIEW

File: appstats.js [hunk 1]
```diff
@@ -219,9 +223,71 @@ class AzureWebAppStats extends AzureAppStats {
 }
 
 class AzureAppInsightStats extends AzureAppStats {
-    constructor(functionNames = []) {
+    constructor(functionNames = [],tokenCredentials,subscriptionId, resourceGroup) {
```
Comment (kkuzmin) [inline appstats.js]: please add spaces after `,`

File: appstats.js [hunk 2]
```diff
@@ -219,9 +223,71 @@ class AzureWebAppStats extends AzureAppStats {
...
+                const query = {
+                    "query": `requests
+                            | where operation_Name =~ '${functionName}'
+                            | order by timestamp desc
+                            | where success == "True" or success == "False"
+                            | summarize errors = countif(success == "True"),
+                                            invocations = countif(success == "True" or success == "False") by operation_Name
+                            | extend details = pack_all()
+                            | summarize Result = make_list(details)
+            `, timespan: 'PT15M'
```
Comment (kkuzmin) [inline appstats.js]: please fix indention in this line

File: appstats.js [hunk 3]
```diff
@@ -219,9 +223,71 @@ class AzureWebAppStats extends AzureAppStats {
...
+                        }
+                    } catch (err) {
+                        return callback(err, null);
+                    }
+                }).catch((err) => {
+                    return callback(err, null);
+                });
+
+            }).catch((err) => {
+                const errMessage = `${err} An error occurred, while getting application insights appId`;
```
Comment (kkuzmin) [inline appstats.js]: I suggest we put `${err}` at the end of the line.

File: appstats.js [hunk 4]
```diff
@@ -219,9 +223,71 @@ class AzureWebAppStats extends AzureAppStats {
...
+                        const data = JSON.parse(result.tables[0].rows[0]);
+                        if (data.length) {
+                            const dataObj = { [data[0].operation_Name]: { invocations: data[0].invocations, errors: data[0].errors } };
+                            return callback(null, dataObj);
+                        } else {
+                            const dataObj = { [functionName]: { invocations: 0, errors: 0 } };
+                            return callback(null, dataObj);
+                        }
+                    } catch (err) {
+                        return callback(err, null);
```
Comment (kkuzmin) [inline appstats.js]: let's add error messages here and below.

File: master.js [hunk 5]
```diff
@@ -148,6 +142,22 @@ class AlAzureMaster {
 
     }
 
+    getTokenCredentials(resource){
+        let credentials={};
+           if (process.env.MSI_ENDPOINT && process.env.MSI_SECRET) {
```
Comment (kkuzmin) [inline master.js]: please fix indention.

File: master.js [hunk 6]
```diff
@@ -120,25 +122,17 @@ class AlAzureMaster {
         this._webAppName = webAppName ? webAppName : process.env.WEBSITE_SITE_NAME;
         this._appFilterJson = process.env.APP_FILTER_JSON ? process.env.APP_FILTER_JSON : '';
         this._appFilterRegex = process.env.APP_FILTER_REGEX ? process.env.APP_FILTER_REGEX : '';
+        
+        //Initialize new SDK
+        this._azureCreds=this.getTokenCredentials();
         if (process.env.FUNCTIONS_EXTENSION_VERSION > '~3') {
-            this._appStats = new AzureAppInsightStats(collectorAzureFunNames);
+            this._appStats = new AzureAppInsightStats(collectorAzureFunNames,this.getTokenCredentials(APPLICATION_INSIGHTS_ENDPOINTS),this._subscriptionId, this._resourceGroup);
```
Comment (kkuzmin) [inline master.js]: Please refactor, the line is too long.
please add spaces between parameters.

File: test/appstats_test.js [hunk 7]
```diff
@@ -283,12 +285,41 @@ describe('App Stats tests', function() {
...
+                ]
+            };
+            var stats = new AzureAppInsightStats(DEFAULT_APP_FUNCTIONS, mockCredentials, process.env.APP_SUBSCRIPTION_ID, process.env.APP_RESOURCE_GROUP);
+            stats.getAppStats('2022-12-22T14:31:39', function (err, appStats) {
+                getInvocationQueryStub.restore();
+                assert.deepEqual(expectedStats, appStats);
+                done();
+            });
+        });
+
```
Comment (kkuzmin) [inline test/appstats_test.js:324]: please add a test where return value from Insights API (`insightsClient.query.execute()`) is mocked to test stats parsing.

**Review by kkuzmin — CHANGES_REQUESTED** [review]

File: appstats.js (hunk 1 above)
Comment (imranalisyed506) [reply appstats.js]: resolved

File: master.js (hunk 6 above)
Comment (imranalisyed506) [reply master.js]: resolved

File: test/appstats_test.js (hunk 7 above)
Comment (imranalisyed506) [reply test/appstats_test.js:324]: resolved

File: appstats.js (hunk 3 above)
Comment (imranalisyed506) [reply appstats.js]: resolved

**Pull request merged**